File: js/src/analytics/correlations.js
/**
 * correlations.js — Pearson correlation, matrix, rolling beta.
 * Spec §4 (Correlations section).
 */

/**
 * Pearson correlation between two same-length series.
 * Returns 0 if either series has zero variance or n < 2.
 * @param {number[]} x
 * @param {number[]} y
 * @returns {number}  in [-1, 1]
 */
export function pearson(x, y) {
    const n = x.length;
    if (n < 2 || y.length < 2) return 0;
    let mx = 0, my = 0;
    for (let i = 0; i < n; i++) { mx += x[i]; my += y[i]; }
    mx /= n; my /= n;
    let cov = 0, vx = 0, vy = 0;
    for (let i = 0; i < n; i++) {
        const dx = x[i] - mx;
        const dy = y[i] - my;
        cov += dx * dy;
        vx  += dx * dx;
        vy  += dy * dy;
    }
    const denom = Math.sqrt(vx * vy);
    if (denom < 1e-14) return 0;
    return cov / denom;
}

/**
 * Correlation matrix for a dict of return series.
 * All series are truncated to the shortest common length.
 * @param {Record<string, number[]>} returnsPerSymbol
 * @returns {{ symbols: string[], matrix: number[][] }}
 */
export function correlationMatrix(returnsPerSymbol) {
    const symbols = Object.keys(returnsPerSymbol);
    const k = symbols.length;
    if (k === 0) return { symbols: [], matrix: [] };

    // Truncate to shortest
    let minLen = Infinity;
    for (const s of symbols) {
        const l = returnsPerSymbol[s].length;
        if (l < minLen) minLen = l;
    }
    if (minLen === Infinity) minLen = 0;

    const series = symbols.map(s => returnsPerSymbol[s].slice(0, minLen));
    const matrix = [];
    for (let i = 0; i < k; i++) {
        const row = new Array(k);
        for (let j = 0; j < k; j++) {
            row[j] = i === j ? 1 : pearson(series[i], series[j]);
        }
        matrix.push(row);
    }
    return { symbols, matrix };
}

/**
 * Rolling beta of target vs baseline over a sliding window.
 * beta_i = cov(target[i-w+1..i], baseline[i-w+1..i]) / var(baseline[i-w+1..i])
 * Returns null for positions before window is full.
 * @param {number[]} target
 * @param {number[]} baseline
 * @param {number}   window
 * @returns {Array<number|null>}
 */
export function rollingBeta(target, baseline, window) {
    const n = target.length;
    const out = new Array(n).fill(null);
    for (let i = window - 1; i < n; i++) {
        const tx = target.slice(i - window + 1, i + 1);
        const bx = baseline.slice(i - window + 1, i + 1);
        let mt = 0, mb = 0;
        for (let j = 0; j < window; j++) { mt += tx[j]; mb += bx[j]; }
        mt /= window; mb /= window;
        let cov = 0, varB = 0;
        for (let j = 0; j < window; j++) {
            cov  += (tx[j] - mt) * (bx[j] - mb);
            varB += (bx[j] - mb) * (bx[j] - mb);
        }
        out[i] = varB < 1e-14 ? 0 : cov / varB;
    }
    return out;
}

File: cpp/src/analytics/correlations.cpp
/**
 * AlgoForge — src/analytics/correlations.cpp
 * Pearson correlation matrix and rolling beta.
 */

#include "core/analytics.hpp"

#include <algorithm>
#include <cmath>
#include <span>
#include <vector>

namespace algoforge::analytics {

double pearson_correlation(
    std::span<const double> x,
    std::span<const double> y
) noexcept {
    const int n = static_cast<int>(std::min(x.size(), y.size()));
    if (n < 2) return 0.0;

    double mx = 0.0, my = 0.0;
    for (int i = 0; i < n; ++i) { mx += x[i]; my += y[i]; }
    mx /= n; my /= n;

    double num = 0.0, sx2 = 0.0, sy2 = 0.0;
    for (int i = 0; i < n; ++i) {
        double dx = x[i] - mx;
        double dy = y[i] - my;
        num += dx * dy;
        sx2 += dx * dx;
        sy2 += dy * dy;
    }

    double denom = std::sqrt(sx2 * sy2);
    return (denom > 1e-15) ? num / denom : 0.0;
}

CorrelationMatrix correlation_matrix(
    const std::unordered_map<std::string, std::vector<double>>& returns_per_symbol
) noexcept {
    CorrelationMatrix cm;
    if (returns_per_symbol.empty()) return cm;

    /* collect symbols in a stable order */
    for (const auto& [sym, _] : returns_per_symbol) {
        cm.symbols.push_back(sym);
    }
    std::sort(cm.symbols.begin(), cm.symbols.end());

    /* find minimum length */
    size_t min_len = SIZE_MAX;
    for (const auto& sym : cm.symbols) {
        min_len = std::min(min_len, returns_per_symbol.at(sym).size());
    }
    if (min_len == SIZE_MAX) min_len = 0;

    const int k = static_cast<int>(cm.symbols.size());
    cm.matrix.assign(k, std::vector<double>(k, 0.0));

    for (int i = 0; i < k; ++i) {
        const auto& vi = returns_per_symbol.at(cm.symbols[i]);
        cm.matrix[i][i] = 1.0;
        for (int j = i + 1; j < k; ++j) {
            const auto& vj = returns_per_symbol.at(cm.symbols[j]);
            std::span<const double> si(vi.data(), min_len);
            std::span<const double> sj(vj.data(), min_len);
            double r = pearson_correlation(si, sj);
            cm.matrix[i][j] = r;
            cm.matrix[j][i] = r;
        }
    }

    return cm;
}

std::vector<double> rolling_beta(
    std::span<const double> target,
    std::span<const double> baseline,
    int                     window
) noexcept {
    const int n = static_cast<int>(std::min(target.size(), baseline.size()));
    if (n < window || window < 2) return {};

    std::vector<double> betas;
    betas.reserve(n - window + 1);

    for (int i = 0; i <= n - window; ++i) {
        std::span<const double> t_win(target.data() + i,   window);
        std::span<const double> b_win(baseline.data() + i, window);

        /* cov(target, baseline) and var(baseline) */
        double mt = 0.0, mb = 0.0;
        for (int j = 0; j < window; ++j) { mt += t_win[j]; mb += b_win[j]; }
        mt /= window; mb /= window;

        double cov = 0.0, var_b = 0.0;
        for (int j = 0; j < window; ++j) {
            cov   += (t_win[j] - mt) * (b_win[j] - mb);
            var_b += (b_win[j] - mb) * (b_win[j] - mb);
        }

        betas.push_back((var_b > 1e-15) ? cov / var_b : 0.0);
    }

    return betas;
}

} /* namespace algoforge::analytics */
